cdap-app-fabric/src/main/java/co/cask/cdap/metadata/MetadataHttpHandler.java

```diff
@@ -246,25 +239,31 @@ public void searchMetadata(HttpRequest request, HttpResponder responder,
     }
   }
 
-  private MetadataEntity getMetadataEntityFromPath(String uri, String suffix) {
+  private MetadataEntity getMetadataEntityFromPath(String uri, String entityType, String suffix) {
     String[] parts = uri.substring((uri.indexOf(Constants.Gateway.API_VERSION_3) +
       Constants.Gateway.API_VERSION_3.length() + 1), uri.lastIndexOf(suffix)).split("/");
     MetadataEntity metadataEntity = new MetadataEntity();
 
     int curIndex = 0;
     while (curIndex < parts.length - 1) {
-      // the api part which we get for program does not have type keyword as part of the uri. It goes like
+      // the api part which we get for program does not have 'type' keyword as part of the uri. It goes like
       // ../apps/appName/programType/ProgramName so handle that correctly
-      if (curIndex >= 2 && parts[curIndex - 2].equalsIgnoreCase("apps")) {
+      if (curIndex >= 2 && parts[curIndex - 2].equalsIgnoreCase("apps") && !parts[curIndex].equalsIgnoreCase
+        ("versions")) {
         metadataEntity = metadataEntity.append(MetadataEntity.TYPE,
                                                ProgramType.valueOfCategoryName(parts[curIndex]).name());
         metadataEntity = metadataEntity.append(MetadataEntity.PROGRAM, parts[curIndex + 1]);
       } else {
-        metadataEntity = metadataEntity.append(API_PART_TO_ENTITY_TYPE.get(parts[curIndex]), parts[curIndex + 1]);
+        if (MetadataClient.ENTITY_TYPE_TO_API_PART.inverse().containsKey(parts[curIndex])) {
+          metadataEntity = metadataEntity.append(MetadataClient.ENTITY_TYPE_TO_API_PART.inverse().get(parts[curIndex]),
+                                                 parts[curIndex + 1]);
+        } else {
+          metadataEntity = metadataEntity.append(parts[curIndex], parts[curIndex + 1]);
+        }
       }
       curIndex += 2;
     }
-    return metadataEntity;
+    return entityType.equalsIgnoreCase("") ? metadataEntity : metadataEntity.changeType(entityType);
```

Comment:
I am not sure I understand the design here. What does it mean to change the type of an entity? Can you explain the rationale behind this design?